modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C

```diff
@@ -190,9 +242,93 @@ ComputeFrictionalForceLMMechanicalContact::incorrectEdgeDroppingPost(
       continue;
 
     _weighted_gap_ptr = &_dof_to_weighted_gap[dof].first;
-    _tangential_vel_ptr = &pr.second;
+    _tangential_vel_ptr[0] = &pr.second[0];
+
+    if (_3d)
+    {
+      _tangential_vel_ptr[1] = &pr.second[1];
+      enforceConstraintOnDof3d(dof);
+    }
+    else
+      enforceConstraintOnDof(dof);
+  }
+}
+
+void
+ComputeFrictionalForceLMMechanicalContact::enforceConstraintOnDof3d(const DofObject * const dof)
+{
+  ComputeWeightedGapLMMechanicalContact::enforceConstraintOnDof(dof);
+
+  // Get normal LM
+  const auto normal_dof_index = dof->dof_number(_sys.number(), _var->number(), 0);
+  const ADReal & weighted_gap = *_weighted_gap_ptr;
+  ADReal contact_pressure = (*_sys.currentSolution())(normal_dof_index);
+  Moose::derivInsert(contact_pressure.derivatives(), normal_dof_index, 1.);
+
+  // Get friction LMs
+  std::array<const ADReal *, 2> & tangential_vel = _tangential_vel_ptr;
+  std::array<dof_id_type, 2> friction_dof_indices;
+  std::array<ADReal, 2> friction_lm_values;
+
+  const unsigned int num_tangents = 2;
+  for (const auto i : make_range(num_tangents))
+  {
+    friction_dof_indices[i] = dof->dof_number(_sys.number(), _friction_vars[i]->number(), 0);
+    friction_lm_values[i] = (*_sys.currentSolution())(friction_dof_indices[i]);
+    Moose::derivInsert(friction_lm_values[i].derivatives(), friction_dof_indices[i], 1.);
+  }
+
+  // Get normalized c and c_t values (if normalization specified
+  const Real c = _normalize_c ? _c / *_normalization_ptr : _c;
+  const Real c_t = _normalize_c ? _c_t / *_normalization_ptr : _c_t;
+
+  ADReal dof_residual;
+  ADReal dof_residual_dir;
 
-    enforceConstraintOnDof(dof);
+  // Primal-dual active set strategy (PDASS)
+  if (contact_pressure < _epsilon)
+  {
+    dof_residual = friction_lm_values[0];
+    dof_residual_dir = friction_lm_values[1];
+  }
+  else
+  {
+    Real epsilon_ad = 1.0e-14;
```

Comment:
Maybe rename to `epsilon_sqrt`? I think the `ad` bit isn't as clear since its type is actually `Real`. Also make `const`